Implement SE-0195, which introduces "Dynamic Member Lookup" Types

lattner opened the pull request:
This is a dusted off and updated version of PR13361,
which switches from DynamicMemberLookupProtocol to @dynamicMemberLookup as
was requested by the final review decision.  This also rebases it,
updates it for other changes in the compiler, improves diagnostics a bit, and adds new tests

lattner committed:
Implement the recently accepted SE-0195 proposal, which introduces "Dynamic
Member Lookup" Types.  This is a dusted off and updated version of PR13361,
which switches from DynamicMemberLookupProtocol to @dynamicMemberLookup as
was requested by the final review decision.  This also rebases it and
updates it for other changes in the compiler.

lattner committed:
Update testcase to remove some unintentional gunk

lattner commented:
@swift-ci please test

lattner commented:
@rudkx I'd appreciate a review whenever you have a chance, thanks!

lattner commented:
@swift-ci please test

lattner committed:
Fix a problem handling infinite superclass chains and update an IDE test I missed.

lattner commented:
@swift-ci please test

lattner commented:
@swift-ci please test

lattner commented:
@swift-ci please test

rudkx commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2863,6 +2863,46 @@ getArgumentLabels(ConstraintSystem &cs, ConstraintLocatorBuilder locator) {
+/// particularly fast in the face of deep class hierarchies or lots of protocol
+/// conformances, but this is fine because it doesn't get invoked in the normal
+/// name lookup path (only when lookup is about to fail).
+static bool isDynamicMemberLookupable(Type ty) {
```
Can I suggest `conformsToDynamicMemberLookup`?

benrimmington replied:
@rudkx But can you *conform* to an attribute?

I suggest `hasDynamicMemberLookup`.

lattner replied:
Thanks, renamed to hasDynamicMemberLookupAttribute

rudkx commented on lib/Sema/OverloadChoice.h:
```diff
@@ -46,6 +46,8 @@ enum class OverloadChoiceKind : int {
   /// \brief The overload choice selects a key path subscripting operation.
   KeyPathApplication,
+  /// \brief The member is looked up through DynamicMemberLookupProtocol.
+  DynamicMemberLookup,
```
Given that this is a separate `OverloadChoiceKind`, it doesn't seem like it's going to interact at all well with `AnyObject` lookup or key paths.

For example it seems like:
```
  var o: AnyObject = ...
  _ = o[dynamicMemberLookup: "property"]
```
will work, if the subscript is marked `@objc`.

What happens if someone attempts `AnyObject` lookup on the property name, though?
```
  _ = o.property
```
Since this feature is about non-ObjC interop, perhaps `@dynamicMemberLookup` attribute checking should only allow subscripts that are not marked `@objc`?

And what happens with key paths?
```
@dynamicMemberLookup
class C {
  subscript(dynamicMember member: String) -> Int { return 7 }
}

_ = \C.testLookup
_ = \C.[dynamicMember: "hi"]
```

lattner replied:
Thank you for the review feedback!

This feature doesn't interact with AnyObject - AnyObject doesn't have the attribute on it, and it doesn't make sense for it to.  Also, this proposal doesn't "take away" direct use of the subscript - there are perfectly reasonable reasons to use it directly as well as through DML.

rudkx replied:
> This feature doesn't interact with AnyObject - AnyObject doesn't have the attribute on it, and it doesn't make sense for it to.

I think some users might look at this from a different direction, especially if they know nothing about implementation details but just view `AnyObject` lookup as magic that can dynamically find things in classes.

```
@dynamicMemberLookup
@objc
class C : NSObject {
  @objc var i: Int
  @objc subscript(dynamicMember member: String) -> Int { ... magic happens ... }
}

var c = C()

c.i      // => c.i
c[dynamicMember: "x"]  // works as expected
c.x     // => c[dynamicMember: "x"]

var o: AnyObject = c
o.i      // => c.i
o[dynamicMember: "x"] // works as expected
o.x    // fails - shouldn't AnyObject-lookup have found my subscript and called it?
```

I'm not suggesting we make this work, but am suggesting that it could save confusion for users if we just outright ban `@objc` on these subscripts.

lattner replied:
That restriction wasn't part of the proposal, and I don't see what adding it would achieve: this has nothing to with Objective-C, the same potential for confusion exists if someone did this:

class X {
  subscript(dynamicMember: ... )
}

@dynamicMemberLookup
class Y : X {
}

Because dynamic lookups would be supported if the static type of a reference is Y but not X.

Also, is a subscript like that even @objc compatible?

rudkx replied:
How about the other issue I mentioned, key paths?

I think users might quite reasonably expect to be able to form key paths with this dynamic lookup behavior, but there is no support here for it.

I strongly suspect we'll crash if we build a key path expression with property components for properties that don't actually exist, so it would be great to have a test or two for that.

rudkx replied:
> this has nothing to with Objective-C, the same potential for confusion exists if someone did this

That's a good point. But your example makes me wonder why we would support finding these subscripts in classes that are not either marked `@dynamicmemberlookup` or inheriting from a class that is.

I don't see a test case like:
```
@dynamicmemberlookup class Base {}
class Derived : Base { subscript(dynamicMember: ...) {} }
_ = Derived().dynamicallyLookedUp
```
and I don't see that mentioned in the SE proposal. Is that supposed to work?

lattner replied:
I just added a testcase for your example:
@dynamicMemberLookup class Base {}

... and the answer is "no" that is not valid. We reject it with this error message:
error: @dynamicMemberLookup attribute requires 'Base' to have a 'subscript(dynamicMember:)' member with a string index

rudkx reviewed (commented):
I mostly focused on the constraint system-related changes and some potential pitfalls, but if you'd like I can take a closer look this week.

lattner committed:
rename isDynamicMemberLookupable -> hasDynamicMemberLookupAttribute
per review feedback.

lattner commented on lib/Sema/CSApply.cpp:
```diff
@@ -2738,23 +2755,59 @@ namespace {
+                              /*trailingClosure*/false,
+                              cs.getConstraintLocator(expr),
+                              /*isImplicit*/false,
+                              AccessSemantics::Ordinary, selected);
```
Yep, looks like a problem with IUO results, great catch, I'll fix.  Thanks!

lattner replied:
Thanks for pointing this out.  IUO problem fixed and testcase added!

rudkx replied:
Thanks!

DougGregor commented on lib/Sema/CSApply.cpp:
```diff
@@ -2738,23 +2755,59 @@ namespace {
+                            loc, /*implicit*/true);
+        auto nameLabel = ctx.getIdentifier("dynamicMember");
+        
+        auto stringType = subscriptDecl->getIndices()->get(0)->getTypeLoc();
```
If the subscript were generic (or in a generic type), `stringType` could end up being a contextual type (i.e., an archetype) within the subscript itself... which isn't what we want. We really need the interface type to be "opened" so the generic type parameters become type variables, and then simplify `stringType` to get the final result. Here's the kind of wacky case I have in mind:

```
subscript<T: ExpressibleByStringLiteral> (dynamicMember name: T) -> MemberLookup<T> { ... }
```

where the context of the member lookup can affect the type used for the incoming string literal.

lattner replied:
Got it, thanks again for your help with this case, I really appreciate it.

DougGregor commented on lib/Sema/CSDiag.cpp:
```diff
@@ -3149,9 +3150,17 @@ void ConstraintSystem::diagnoseAssignmentFailure(Expr *dest, Type destTy,
+    // access. Emit a more specific diagnostic.
+    if (sub->getIndex()->isImplicit() &&
+        sub->getArgumentLabels().size() == 1 &&
+        sub->getArgumentLabels().front().str() == "dynamicMember")
```
Maybe add `dynamicMember` to `KnownIdentifiers.def`, so you don't have to hardcode the string repeatedly?

lattner replied:
I can do that if you'd prefer, but is it better in practice?  We'd still be hardcoding the string repeatedly, just in a different way/place.

lattner replied:
I'm happy to do it, and not arguing, just seek to understand why it is better.

DougGregor replied:
Mostly I think it's an anti-pattern to have the same string literal repeated in many places. Even a #define would be better, but since you're always working with identifiers, `KnownIdentifiers.def` seems the best place.

DougGregor commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -918,6 +921,78 @@ static bool isRelaxedIBAction(TypeChecker &TC) {
+  // Lookup the implementations of our subscript.
+  auto candidates = TC.lookupMember(decl, type, subscriptName, lookupOptions);
+  
+  // If we have none, then there is no conformance.
```
It's not longer a conformance.

lattner replied:
Fixed, thanks!

DougGregor commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -918,6 +921,78 @@ static bool isRelaxedIBAction(TypeChecker &TC) {
+ERROR(invalid_retroactive_conformance_dmlp,none,
+      "retroactive conformance of %0 to 'DynamicMemberLookupProtocol' is not "
+      "allowed, only primary type declarations may conform", (Type))
+#endif
```
Development cruft.

lattner replied:
Fixed, thanks!

DougGregor commented on lib/Sema/TypeChecker.h:
```diff
@@ -2551,6 +2551,13 @@ class EncodedDiagnosticMessage {
 };
 
+/// Given a subscript defined as "subscript(dynamicMember:)->T", return true if
+/// it is an acceptable implementation of the DynamicMemberLookupProtocol
```
`dynamicMemberLookup` attribute, now

lattner replied:
👍

DougGregor commented on test/NameBinding/dynamic-member-lookup.swift:
```diff
@@ -0,0 +1,231 @@
+
+func testMutableExistential(a : PyVal, b : MyType) -> PyVal {
+  a.x.y = b
+  b.x.y = b
```
This is hitting the narrow case where an existential is represented as a nominal type (which is why the test succeeds now); I'd expect that, e.g., `AnyObject & PyVal` would fail because that's represented as a protocol composition type.

lattner replied:
Fixed, thank you again for your testcase.

DougGregor commented on test/NameBinding/dynamic-member-lookup.swift:
```diff
@@ -0,0 +1,231 @@
+  }
+}
+class DerivedClass : BaseClass {
+}
```
Fun case to try: a derived class overrides `subscript(dynamicMember:)` to make it settable.

lattner replied:
Works 👍, testcase added.

DougGregor commented on test/NameBinding/dynamic-member-lookup.swift:
```diff
@@ -0,0 +1,231 @@
+var global = 42
+
+@dynamicMemberLookup
+struct Gettable {
```
It's important to test `subscript(dynamicMember:)` within generic types and probably as a generic subscript.

lattner replied:
👍 Thank you for the test cases, added.

DougGregor commented:
It's getting there! My primary concerns with the implementation right now are:
* Support for `subscript(dynamicMember:)` within generics and that is generic, because the handling of index types looks like it'll be incorrect.
* Clarification on existential types (and type parameters, I guess), which I think should be supported but aren't because we're only considering nominal types.

NachoSoto commented:
_From the peanut gallery_: given the potential for this to introduce regressions, would it make sense to make sure that the source-compatibility suite is up-and-running correctly? [I counted 20 targets that are currently ignored](https://github.com/apple/swift-source-compat-suite/blob/master/projects.json). For example, because [SR-6658](https://bugs.swift.org/browse/SR-6658) wasn't fixed on time, there was [a regression on `Swift 4.1`](https://bugs.swift.org/browse/SR-6796) that could have been caught otherwise, and I'm worried that as time goes on, there could be more of these.

lattner commented:
@DougGregor Thanks for the feedback.  Can you give more detailed examples of the cases you're concerned about?  It isn't immediately obvious to me what cases you're concerned about, but I'm happy to make sure they work (or get fixed) if I can get the (sketch of) a testcase.

@NachoSoto This isn't particularly likely to introduce regressions since it is an additive feature.  It is much more likely that there will be some new thing that should be supported by it that isn't already.

DougGregor commented:
@NachoSoto, we aren't going to hold contributions hostage because of unrelated source-compatibility failures. I'm happy to engage on a discussion of [SR-6658](https://bugs.swift.org/browse/SR-6658) and [SR-6797](https://bugs.swift.org/browse/SR-6796) in an appropriate forum, but on a pull request for a separate feature that was approved by the Swift Evolution process, it's noise.

DougGregor commented:
@lattner , so test case sketches:

Adding a setter for the dynamicMember subscript in a derived class:
```
class DerivedClassWithSetter : BaseClass {
  override subscript(dynamicMember member: String) -> Int {
    get { return super[dynamicMember: member] }
    set { }
   }
}
```

Generics:

```
@dynamicMemberLookup
 struct SettableGeneric1<T> {
   subscript(dynamicMember member: StaticString) -> T? {
     get {return nil}
     set {}
   }
 }

@dynamicMemberLookup
 struct SettableGeneric2<T> {
   subscript<U: ExpressibleByStringLiteral>(dynamicMember member: U) -> T? {
     get {return nil}
     set {}
   }
 }
```

More existentials:

```
protocol SubPyVal : PyVal { }

func testMutableExistential(a : AnyObject & PyVal, b : SubPyVal)  {
 a.x.y = b
 b.x.y = b
}
```

More generics:

```
func testGeneric<T: PyVal, U: SubPyVal>(a: T, b: U) {
 a.x.y = b
 b.x.y = b
}
```

lattner committed:
Fix a bozo bug pointed out by Paul Hudson where we were adding DML results
even when we got normal results.

lattner committed:
Fix another problem reported by Paul Hudson, were we dropped substitutions
on a ConcreteDeclRefExpr because we were using the wrong locator.

lattner committed:
Fix problem passing the wrong locator when dealing with IUO results,
thanks to Mark Lacey for pointing out this case.

lattner committed:
add a testcase requestd by Mark.

lattner committed:
fix a problem handling protocol compositions and protocol refinements,
add a testcase that already works for derived classes overriding the
subscript.

Thanks to Doug Gregor for both testcases!

lattner committed:
Fix a problem where we didn't default the index type of a dynamicMemberLookup
to String because there was no LiteralConformsTo constraint.  Also fix a problem
where we'd dig the incorrect type out of a decl, applying an archetype to the
resolved string type.  Add testcases to cover some generics cases.

Many thanks to Doug Gregor for the testcases as well as helping me out with one
of the tricky parts of this patch!

lattner commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2863,6 +2863,46 @@ getArgumentLabels(ConstraintSystem &cs, ConstraintLocatorBuilder locator) {
+/// @dynamicMemberLookup attribute on it.  This implementation is not
+/// particularly fast in the face of deep class hierarchies or lots of protocol
+/// conformances, but this is fine because it doesn't get invoked in the normal
+/// name lookup path (only when lookup is about to fail).
```
Ok, I'm happy to add that cache.  Do you recommend adding this as a bit to TypeBase, or somewhere else?

DougGregor replied:
I think it'd be fine to make it a `DenseMap` in the `ConstraintSystem` itself. My concern is about particular constraint system going exponential (or just having a large solution space) and this to kick in along all of the failing paths.

lattner replied:
SGTM, cache added. thanks!

lattner commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -918,6 +921,78 @@ static bool isRelaxedIBAction(TypeChecker &TC) {
+  TC.Context.getProtocol(KnownProtocolKind::ExpressibleByStringLiteral);
+  
+  return indices->size() == 1 &&
+  !indices->get(0)->isVariadic() &&
```
oooh, indeed, not sure what happened there.  Fixed, thanks for noticing that!

lattner committed:
more cleanups caught by Doug.

lattner commented:
Awesome, thank you for the detailed feedback Doug!  I added all the test cases and I think I addressed all of your comments above.

lattner commented:
The two open issues is "where to cache the bit" and "do you really want this in knownIdentifiers.def"?  I also owe @rudkx and answer about key paths.

lattner committed:
Change "dynamicMember" magic string to being in known identifiers.

lattner commented:
I tried caching the bit in RecursiveTypeProperties which seems natural.  The problem with that is that types are formed before name lookup is done.  Instead I will actually listen to Doug for once ;-) and do what he says: try a map in ConstraintSystem.

lattner committed:
Add a cache for the 'isDynamicMemberLookup' query to ConstraintSystem so a type
doesn't get recomputed as the system does backtracking.

lattner committed:
Implement keypaths for @dynamicMemberLookup results by desugaring them
to the same form as if the user wrote the subscript out manually.

lattner commented:
@swift-ci please test

lattner commented:
@swift-ci please test

lattner commented:
@swift-ci please test

lattner commented:
@swift-ci please test

lattner commented:
Keypath support is now implemented.

@rudkx and @DougGregor, I think I have addressed all of the concerns you've raised upthread.  Thank you for the detailed feedback, I really appreciate your effort on this.  I just kicked off a test suite check, please take another look whenever you have time and I'd like to merge it if possible.  Thanks!!

lattner commented:
@rudkx and @DougGregor, do you have any concerns with me hitting 'squash and merge'?  I'm happy to continue improving this in subsequent PR's if anything new pops up.  Thanks!

DougGregor commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2869,47 +2869,55 @@ getArgumentLabels(ConstraintSystem &cs, ConstraintLocatorBuilder locator) {
-    return false;
-  }
+static bool hasDynamicMemberLookupAttribute(Type ty,
+                       llvm::DenseMap<Type, bool> &IsDynamicMemberLookupCache) {
```
I suggest making the input a `CanType`, and I guess (annoyingly) we shouldn't cache anything is `ty->hasTypeVariable()`. Sorry for not making this clearer earlier.

DougGregor commented:
LGTM!

lattner commented:
Thank you!

rudkx reviewed (commented):
Looks good, thanks!

lattner commented:
Thanks @DougGregor, I'm improving the caching thing in PR 14697.
